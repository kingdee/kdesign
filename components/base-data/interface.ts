import { tuple } from '../_utils/type'

export const Modes = tuple('multiple', 'single')
export type Mode = typeof Modes[number]

export type SelectValue = string | string[] | number | number[] | IOptionsProps[] | null

export interface IAdvancedSelectorProps<T = SelectValue> {
  prefixCls?: string // 样式前缀
  className?: string // 样式名
  children?: React.ReactNode
  style?: React.CSSProperties // 内联样式
  disabled?: boolean // 禁用
  dropdownStyle?: React.CSSProperties // 下拉样式
  notFoundContent?: React.ReactNode // 列表为空时文本
  placeholder?: string // 选择框默认文字
  showFrequent?: boolean // 是否显示历史记录与我的收藏面板
  showCollectIcon?: boolean // 是否显示收藏按钮
  showDetailIcon?: boolean // 是否显示详情按钮
  suffixIcon?: React.ReactNode // 选择框后缀图标
  mode?: Mode // 模式
  value?: IOptionsProps[] | IOptionsProps // 指定选中的条目
  columns?: [] // 下拉列表的展示列
  options?: [] // 下拉列表的数据
  size?: number // 下拉列表显示数量
  optionLabelProp?: string // 回填到选择框的属性
  dropdownFooterRender?: React.ReactNode // 自定义下拉框footer内容
  visible?: boolean
  collectList?: [] // 收藏列表
  historyList?: [] // 我的历史
  loading?: boolean // 加载中状态
  searchField?: string // 搜索字段 默认 all（全部字段）
  onChange?: () => void // 选中option调用此函数
  onSearch?: (val: string) => void // input的value变化时调用
  onSelect?: (value: T extends (infer I)[] ? I : T, option: React.ReactElement<any>) => void
  onShowDetail?: (value: IOptionsProps[]) => void
  onShowMore?: () => void
  onCollect?: (flag: boolean, value: IOptionsProps) => void
}

export interface IIndxPosListProps {
  end: number
  start: number
}

export interface IColumnsProps {
  title?: string
  key: string
}

export interface IOptionsProps {
  value?: string | number
  label?: string
}

export interface ISearchInfoProps {
  searchValue?: string
  previousInputValue: string
  previousEditValue?: string
  searchIndex?: number
  selectionStart?: number
  deleteEndIndx: number
  editOptions?: any[]
}
