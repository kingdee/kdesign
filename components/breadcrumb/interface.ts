import { ReactNode } from 'react'
import { DropDownProps } from '../dropdown'
import { tuple } from '../_utils/type'

export const ColorModelTypes = tuple('emphasize', 'weaken')
export type ColorModelType = typeof ColorModelTypes[number]

export interface IBreadcrumbProps {
  style?: Record<string, unknown> // 内联样式
  className?: string // 类名
  colorModel?: ColorModelType // 颜色模式，默认是 'emphasize'
  prefixCls?: string // 样式前缀
  items: Array<IBreadcrumbItem> // 面包屑导航的各项，支持传入 ReactNode 类型
  separator?: ReactNode // 分隔符，默认是 '/'
  children?: ReactNode // 子元素
  onItemClick?: (item: IBreadcrumbItem, index: number) => void // 点击面包屑导航的回调函数
}

export interface IBreadcrumbItems {
  item: IBreadcrumbItem
  index: number
  separator?: ReactNode
  openEllipsis?: boolean
  onItemClick?: (item: IBreadcrumbItem, index: number) => void // 点击面包屑导航的回调函数
}

export interface IBreadcrumbItem {
  className?: string // 类名
  title?: ReactNode
  dropdownProps?: DropDownProps
  href?: string // 链接地址，不能和path共用
  path?: string // 拼接地址，不能和href共用
  icon?: ReactNode
}

export interface IItemsWidth {
  width: number
  index: number
}
