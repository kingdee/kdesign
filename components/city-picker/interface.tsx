import * as React from 'react'
import { tuple } from '../_utils/type'

export type City = {
  id?: string | number
  country?: string
  province?: string
  name?: string
  [key: string]: any
}

export const BorderTypes = tuple('none', 'underline', 'bordered')
export type BorderType = typeof BorderTypes[number]

export const Sizes = tuple('large', 'middle', 'small') // 选择框大小
export type Size = typeof Sizes[number]

export type Type = 'domestic' | 'foreign'

export type CityList = Array<City>

export interface CityPickerProps {
  prefixCls?: string // 样式前缀
  size?: Size // 尺寸
  borderType?: BorderType // 边框类型
  notFoundContent?: React.ReactNode // 列表为空时文本
  showArrow?: boolean // 显示下拉小箭头
  defaultOpen?: boolean // 默认是否展开下拉列表
  allowClear?: boolean // 允许清除内容
  placeholder?: React.ReactNode // 选择框默认文字
  type?: Type
  loading?: boolean
  domesticList?: CityList
  foreignList?: CityList
  commonList?: CityList
  className?: string
  disabled?: boolean
  style?: React.CSSProperties
  title?: string
  clearIcon?: React.ReactNode // 清除图标
  suffixIcon?: React.ReactNode // 选择框后缀图标
  optionHighlightProps?: string // 搜索时 根据该属性值高亮城市名
  value?: string | number
  defaultValue?: string | number
  description?: React.ReactNode
  showDescription?: boolean
  onChange?: (value: string | number, city: City) => void
  onBlur?: (e: React.ChangeEvent<HTMLSpanElement>) => void
  onFocus?: () => void
  onSearch?: (value: string) => void
  onClear?: () => void
  itemRender?: (city: City) => React.ReactNode
  onTabPaneChange?: (id: Type) => void
}

export interface ICityPickerOptionProps {
  disabled?: boolean
  city?: City
  value?: string | number
  children?: React.ReactNode
  className?: string
  style?: React.CSSProperties
  onChangeSelect?: (city: City | null) => void
  renderCityInfo?: (city: City, flag?: boolean, symbol?: string) => React.ReactNode
  itemRender?: (city: City) => React.ReactNode
}
