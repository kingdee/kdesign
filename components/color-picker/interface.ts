import { tuple } from '../_utils/type'
import { PopperProps } from '../_utils/usePopper'

export const ColorTypes = tuple('HEX', 'HSB', 'RGB', 'HSL', 'themeColor')
export type ColorType = typeof ColorTypes[number]
export interface IColorTypesObj {
  type: ColorType
  value: string
}

export const BorderTypes = tuple('underline', 'bordered')
export type BorderType = typeof BorderTypes[number]
export interface IColorPickerProps extends PopperProps {
  // 设置颜色选择器的值，可以是十六进制颜色值，RGB颜色值，HSL颜色值，或者颜色名称
  value: string
  // 设置颜色选择器的默认值，可以是十六进制颜色值，RGB颜色值，HSL颜色值，或者颜色名称
  defaultValue: string
  // 选择器输入框类名
  className: string
  // 选择输入框样式
  style: React.CSSProperties
  // 输入框类型
  borderType: BorderType
  // 输入框右侧的图标
  suffixIcon: (rgbColor: string) => React.ReactNode
  // 功能色名称
  functionalColorName: string
  // 功能色色值
  functionalColor: string
  // 功能色开关的名称
  switchName: { name: string; internationalName?: string }
  // 是否跟随功能色,定义了功能色才能控制显示跟随主题色开关
  showSwitch: boolean
  // 是否展示色彩格式转换框
  showColorTransfer: boolean
  // 是否展示系统预设颜色
  showPresetColor: boolean
  // 占位字符
  placeholder: string
  // 用户自定义的预设颜色
  presetColor: string[]
  // 是否展示拾色容器
  showColorPickerBox: { showBox: boolean; showHue: boolean; showOpacity: boolean }
  // 颜色选择面板显隐配置
  defaultOpen: boolean
  // 当颜色选择器的值发生变化时触发的回调函数
  onChange: (inputValue: string, formatColorArr: Array<IColorTypesObj>) => void
}

export interface IColorPickerPanelProps {
  alpha: number
  alphaNoVerifyVal: string
  clickedColorIndex?: number
  colTypeArr: Array<IColorTypesObj>
  correctColorValue: string
  currentColorType: IColorTypesObj['type']
  functionalColor?: string
  functionalColorName: string
  switchName: { name: string; internationalName?: string }
  inputRef: any
  isFollow: boolean
  presetColor?: string[]
  showSwitch?: boolean
  showColorTransfer?: boolean
  showPresetColor?: boolean
  showColorPickerBox?: { showBox?: boolean; showHue?: boolean; showOpacity?: boolean }
  showColorPickerPanel: boolean
  value: string
  setCurrentColorType: (currentColorType: IColorTypesObj['type']) => void
  setColTypeArr: (colTypeArr: Array<IColorTypesObj>) => void
  setClickedColorIndex: (clickedColorIndex: number) => void
  setIsFollow: (isFollow: boolean) => void
  setAlphaNoVerifyVal: (NoVerifyVal: string) => void
  setAlpha: (alpha: number) => void
  setCorrectColorValue: (colorValue: string) => void
  setInputColorValue: (colorValue: string) => void
  setShowPanel: (showPanel: boolean) => void
  onChange?: (inputValue: string, formatColorArr: Array<IColorTypesObj>) => void
}
