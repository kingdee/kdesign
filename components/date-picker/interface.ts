import { BorderType, InputSiteType } from '../input/input'
import { DisabledDataProps } from './date-picker'

export type WeekTitleType = [string, string, string, string, string, string, string]
export type MonthTitleType = [
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
]

export interface Locale {
  // locale: string
  'DatePicker.placeholder': string
  'DatePicker.yearPlaceholder': string
  'DatePicker.quarterPlaceholder': string
  'DatePicker.monthPlaceholder': string
  'DatePicker.weekPlaceholder': string
  'DatePicker.timePlaceholder': string
  'DatePicker.now': string
  'DatePicker.confrim': string
  'DatePicker.today': string
  'DatePicker.year': string
  'DatePicker.month': string
  'DatePicker.weekTitle': WeekTitleType
  'DatePicker.monthTitle': MonthTitleType
  'DatePicker.rangePlaceholder': [string, string]
  'DatePicker.rangeYearPlaceholder': [string, string]
  'DatePicker.rangeMonthPlaceholder': [string, string]
  'DatePicker.rangeWeekPlaceholder': [string, string]
  'DatePicker.rangeQuarterPlaceholder': [string, string]
  'DatePicker.rangeTimePlaceholder': [string, string]
}

export interface InnerLocale {
  locale: string
  placeholder: string
  yearPlaceholder: string
  quarterPlaceholder: string
  monthPlaceholder: string
  weekPlaceholder: string
  timePlaceholder: string
  now: string
  confrim: string
  today: string
  year: string
  month: string
  weekTitle: WeekTitleType
  monthTitle: MonthTitleType
  rangePlaceholder: [string, string]
  rangeYearPlaceholder: [string, string]
  rangeMonthPlaceholder: [string, string]
  rangeWeekPlaceholder: [string, string]
  rangeQuarterPlaceholder: [string, string]
  rangeTimePlaceholder: [string, string]
}

export type InnerLocaleKey = keyof InnerLocale

export type PanelMode = 'time' | 'date' | 'week' | 'month' | 'quarter' | 'year' | 'decade'

export type PickerMode = Exclude<PanelMode, 'datetime' | 'decade'>

export type DateType = Date

export type CustomFormat = (value: DateType) => string

export type DateFormat = string

export type DateFormats = Array<string>

export type NullableDateType = DateType | null | undefined

export type OnSelect = (value: DateType, type: 'key' | 'mouse' | 'submit') => void
export interface DisabledTimes {
  disabledHours?: () => number[]
  disabledMinutes?: (hour: number) => number[]
  disabledSeconds?: (hour: number, minute: number) => number[]
}

export type DisabledTime = (date: DateType | null) => DisabledTimes

export type OnPanelChange = (value: DateType, mode: PanelMode) => void

export type EventValue = DateType | null
export type RangeValue = [EventValue, EventValue] | null

export type RangeList = {
  label: string
  onClick: () => void
  onMouseEnter: () => void
  onMouseLeave: () => void
}[]

export interface Components {
  button?: React.ComponentType | string
  rangeItem?: React.ComponentType | string
}

export interface PanelSharedProps {
  value?: NullableDateType
  defaultPickerValue?: DateType
  locale: InnerLocale
  disabledDate?: DisabledDataProps

  prevIcon?: React.ReactNode
  nextIcon?: React.ReactNode
  superPrevIcon?: React.ReactNode
  superNextIcon?: React.ReactNode

  // 面板操作（焦点操作、键盘操作）next version
  // operationRef: React.MutableRefObject<PanelRefProps>

  // onPanelChange: (mode: PanelMode | null, viewValue: DateType) => void
}

export type DateRender = (currentDate: DateType, today: DateType) => React.ReactNode

export type CellRender = (currentDate: DateType) => React.ReactNode

export type RangeType = 'start' | 'end'

export interface RangeInfo {
  range: RangeType
}

export type RangeDateRender = (currentDate: DateType, today: DateType, info: RangeInfo) => React.ReactNode

export interface SharedTimeProps extends DisabledTimes {
  format?: string
  showNow?: boolean
  showHour?: boolean
  showMinute?: boolean
  showSecond?: boolean
  use12Hours?: boolean
  hourStep?: number
  minuteStep?: number
  secondStep?: number
  hideDisabledOptions?: boolean
  defaultValue?: DateType
}

export interface OutInputCommonProps {
  format: DateFormat
  picker: PickerMode
  id?: string
  autoFocus?: boolean
  allowClear?: boolean
  className?: string
  borderType?: BorderType
  style?: React.CSSProperties
  size?: InputSiteType
  readOnly?: boolean
  suffixIcon?: React.ReactNode
  clearIcon?: React.ReactNode

  onFocus?: React.FocusEventHandler<HTMLInputElement>
  onBlur?: React.FocusEventHandler<HTMLInputElement>
  onMouseDown?: React.MouseEventHandler<HTMLDivElement>
  onMouseUp?: React.MouseEventHandler<HTMLDivElement>
  onMouseEnter?: React.MouseEventHandler<HTMLDivElement>
  onMouseLeave?: React.MouseEventHandler<HTMLDivElement>
  onClick?: React.MouseEventHandler<HTMLDivElement>
  onContextMenu?: React.MouseEventHandler<HTMLDivElement>
}

export interface InputCommonProps {
  panelDivRef: React.RefObject<HTMLDivElement>
  needConfirmButton: boolean
  open: boolean
  prefixCls: string
  locale: InnerLocale
  dataOrAriaProps?: Record<string, any>
  status?: 'error'
}

export interface OutPopperProps {
  dropdownClassName?: string
  popupStyle?: React.CSSProperties
  showArrow?: boolean
}

export interface TimeUnit {
  label: React.ReactText
  value: number
  disabled: boolean
}
