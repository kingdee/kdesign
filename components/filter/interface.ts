import React from 'react'
import { Menu } from '../dropdown/menu'
import { IQuickSearchProps } from '../search'
import { DirectionType } from '../config-provider/ConfigContext'

type TabKey = 'condition' | 'scheme'

export type TFilterValue = {
  [key: string]: string[] | IScheme
}

export type TOption = {
  value?: string
  label?: string
  component?: React.ReactNode
}

export interface ICondition {
  key: string
  label: string
  options: TOption[]
  required?: boolean
  multiple?: boolean
  addition?: React.ReactNode
}

export interface IField {
  name: string
  label: string
  required?: boolean
  operators: React.ReactElement
  assignOptions: Record<string, React.ReactElement>
}

export interface ISchemeItem {
  fieldName: string
  value: string
  operator: string
  required?: boolean
}

export interface IScheme {
  name?: string
  preset?: boolean
  items: ISchemeItem[]
}

export interface IFilterProps {
  spread?: boolean
  fields?: IField[]
  className?: string
  schemes?: IScheme[]
  value?: TFilterValue
  defaultTabKey?: TabKey
  defaultSpread?: boolean
  title?: React.ReactNode
  conditions?: ICondition[]
  search?: IQuickSearchProps
  style?: React.CSSProperties
  defaultValue?: TFilterValue
  onSchemeSave?: (scheme: IScheme) => void
  onSpreadChange?: (spread: boolean) => void
  getSchemeSettingMenu?: (scheme: IScheme) => typeof Menu
  onChange?: (value: TFilterValue, condition?: ICondition, option?: TOption) => void
}

export interface IConditionProps {
  tab: string
  prefixCls: string
  value: TFilterValue
  conditions: ICondition[]
  FilterLangMsg: Record<string, any>
  onConditionRemove: (key: string, e: React.MouseEvent) => void
  onConditionChange: (value: TFilterValue, condition?: ICondition, option?: TOption) => void
}

export interface ISchemeProps {
  tab: string
  fields: IField[]
  prefixCls: string
  schemes: IScheme[]
  direction: DirectionType
  value: { scheme: IScheme }
  FilterLangMsg: Record<string, any>
  onSchemeSave: (scheme: IScheme) => void
  onSchemeSearch: (scheme: IScheme) => void
  getSchemeSettingMenu?: (scheme: IScheme) => void
}
