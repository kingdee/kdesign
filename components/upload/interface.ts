import * as React from 'react'
import { ProgressProps } from '../progress'

export type UploadFileStatus = 'error' | 'success' | 'done' | 'uploading' | 'removed' | 'notStart'
export type UploadRequestMethod = 'POST' | 'PUT' | 'PATCH' | 'post' | 'put' | 'patch'

export type Action = string | ((file: UploadFile) => string) | ((file: UploadFile) => PromiseLike<string>)

export interface HttpRequestHeader {
  [key: string]: string
}

export interface UploadFile<T = any> extends File {
  uid: string
  size: number
  name: string
  type: string
  fileName?: string
  readonly lastModified: number
  readonly lastModifiedDate: Date
  webkitRelativePath: string
  url?: string
  status?: UploadFileStatus
  percent?: number
  thumbUrl?: string
  originFileObj?: File | Blob
  response?: T
  error?: any
  linkProps?: any
  xhr?: T
  preview?: string
}

export interface UploadChangeParam<T extends UploadFile> {
  file: T
  fileList: Array<UploadFile>
  event?: { percent: number }
}

export interface ShowUploadListInterface {
  showRemoveIcon?: boolean
  showPreviewIcon?: boolean
  showDownloadIcon?: boolean
  removeIcon?: React.ReactNode | ((file: UploadFile) => React.ReactNode)
  downloadIcon?: React.ReactNode | ((file: UploadFile) => React.ReactNode)
}

export interface UploadLocale {
  uploading?: string
  removeFile?: string
  downloadFile?: string
  uploadError?: string
  previewFile?: string
}

export type UploadType = 'drag' | 'select'
export type UploadListType = 'text' | 'picture'
export type UploadListProgressProps = Omit<ProgressProps, 'percent' | 'type'>

export type ItemRender<T = any> = (
  originNode: React.ReactElement,
  file: UploadFile,
  fileList?: Array<UploadFile<T>>,
) => React.ReactNode

type PreviewFileHandler = (file: File | Blob) => PromiseLike<string>
type TransformFileHandler = (file: UploadFile) => string | Blob | File | PromiseLike<string | Blob | File>

export interface UploadProps<T = any> {
  type?: UploadType
  name?: string
  defaultFileList?: Array<UploadFile<T>>
  fileList?: Array<UploadFile<T>>
  action?: Action
  directory?: boolean
  data?: Record<string, unknown> | ((file: UploadFile<T>) => Record<string, unknown>)
  method?: 'POST' | 'PUT' | 'PATCH' | 'post' | 'put' | 'patch'
  headers?: HttpRequestHeader
  showUploadList?: boolean | ShowUploadListInterface
  multiple?: boolean
  accept?: string
  beforeUpload?: (file: UploadFile, FileList: UploadFile[]) => boolean | Promise<void | Blob | File>
  onStart?: (file: UploadFile<T>) => void
  onChange?: (info: UploadChangeParam<UploadFile>) => void
  listType?: UploadListType
  className?: string
  onPreview?: (file: UploadFile<T>) => void
  onDownload?: (file: UploadFile<T>) => void
  onRemove?: (file: UploadFile<T>) => void | boolean | Promise<void | boolean>
  supportServerRender?: boolean
  style?: React.CSSProperties
  disabled?: boolean
  prefixCls?: string
  customRequest?: (options: UploadRequestOption) => void
  withCredentials?: boolean
  openFileDialogOnClick?: boolean
  locale?: UploadLocale
  id?: string
  previewFile?: PreviewFileHandler
  /** @deprecated Please use `beforeUpload` directly */
  transformFile?: TransformFileHandler
  iconRender?: (file: UploadFile<T>, listType?: UploadListType) => React.ReactNode
  isImageUrl?: (file: UploadFile) => boolean
  progress?: UploadListProgressProps
  itemRender?: ItemRender<T>
  onError?: (error: Error, ret: Record<string, unknown>, file: UploadFile) => void
  onSuccess?: (response: Record<string, unknown>, file: UploadFile, xhr: XMLHttpRequest) => void
  onProgress?: (event: UploadProgressEvent, file: UploadFile) => void
  onMouseEnter?: (e: React.MouseEvent<HTMLDivElement>) => void
  onMouseLeave?: (e: React.MouseEvent<HTMLDivElement>) => void
  onClick?: (e: React.MouseEvent<HTMLDivElement> | React.KeyboardEvent<HTMLDivElement>) => void
}

export interface UploadState<T = any> {
  fileList: UploadFile<T>[]
  dragState: string
}

export interface UploadListProps<T = any> {
  listType?: UploadListType
  onPreview?: (file: UploadFile<T>) => void
  onDownload?: (file: UploadFile<T>) => void
  onRemove?: (file: UploadFile<T>) => void | boolean
  items?: Array<UploadFile<T>>
  progress?: UploadListProgressProps
  prefixCls?: string
  showRemoveIcon?: boolean
  showDownloadIcon?: boolean
  showPreviewIcon?: boolean
  removeIcon?: React.ReactNode | ((file: UploadFile) => React.ReactNode)
  downloadIcon?: React.ReactNode | ((file: UploadFile) => React.ReactNode)
  locale: UploadLocale
  previewFile?: PreviewFileHandler
  iconRender?: (file: UploadFile<T>, listType?: UploadListType) => React.ReactNode
  isImageUrl?: (file: UploadFile) => boolean
  appendAction?: React.ReactNode
  itemRender?: ItemRender<T>
}

export interface UploadRequestError extends Error {
  status?: number
  method?: UploadRequestMethod
  url?: string
}

export interface UploadProgressEvent extends ProgressEvent {
  percent: number
}

export interface UploadRequestOption<T = any> {
  onProgress?: (event: UploadProgressEvent) => void
  onError?: (event: UploadRequestError | ProgressEvent, body?: T) => void
  onSuccess?: (body: T, xhr: XMLHttpRequest) => void
  data?: Record<string, unknown>
  filename?: string
  file: UploadFile
  withCredentials?: boolean
  action: string
  headers?: HttpRequestHeader
  method: UploadRequestMethod
}

export interface IFileItem {
  file: UploadFile
  prefixCls: string
  listType: string
  disabled: boolean
  onPreview?: (file: UploadFile) => void
  handleReUpload: (file: UploadFile, e: React.MouseEvent) => void
  handleRemove: (file: UploadFile, e: React.MouseEvent) => void
}
