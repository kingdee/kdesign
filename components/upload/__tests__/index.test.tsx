import React from 'react'
import { render, mount } from 'enzyme'
import Upload from '../index'
import Icon from '../../icon'
import Button from '../../button'
import { UploadListType, UploadProps, UploadFile } from '../interface'
import mountTest from '../../../tests/shared/mountTest'

const sleep = (timeout = 0) => new Promise((resolve) => setTimeout(resolve, timeout))
const dragButton = (
  <div>
    <Icon type="add" style={{ fontSize: 16, color: '#666', fontWeight: 'bolder' }} />
    <div>点击或拖拽上传</div>
  </div>
)
const uploadButton = <Button icon={<Icon type="upload" />}>Upload</Button>
const defaultFileList = [
  {
    uid: '1',
    size: '100',
    status: 'done',
    name: 'xxx.png',
    response: 'Server Error 500',
    url: 'http://www.baidu.com/xxx.png',
  },
  {
    uid: '2',
    size: '100',
    name: 'yyy.png',
    status: 'done',
    url: 'http://www.baidu.com/yyy.png',
  },
  {
    uid: '3',
    size: '100',
    name: 'zzz.png',
    status: 'error',
    response: 'Server Error 500',
    url: 'http://www.baidu.com/zzz.png',
  },
] as unknown as UploadFile[]
const action = 'https://www.mocky.io/v2/5cc8019d300000980a055e76'
const defaultUploadProps: UploadProps = {
  defaultFileList,
  action,
}

describe('Upload', () => {
  // 1.mount test
  mountTest(Upload)

  // 2.render test
  it('renders correctly', () => {
    const wrapper = render(<Upload.Dragger>{dragButton}</Upload.Dragger>)
    expect(wrapper).toMatchSnapshot()
    const types = ['text', 'picture']
    types.forEach((type: UploadListType) => {
      const wrapper = render(<Upload listType={type}></Upload>)
      expect(wrapper).toMatchSnapshot()
    })
  })

  // 3.render no child without errors
  it('render no child without errors', () => {
    expect(mount(<Upload></Upload>)).toMatchSnapshot()
  })

  // 4. render null or undefined without errors
  it('render null or undefined without errors', () => {
    expect(
      mount(
        <Upload>
          {null}
          {undefined}
        </Upload>,
      ),
    ).toMatchSnapshot()
  })

  // 5.displayName
  it('should have displayName static property', () => {
    const wrapper = mount(<Upload></Upload>)
    expect((wrapper.type() as any).displayName).toBe('Upload')
  })

  // 6.render defaultFileList without error
  it('render defaultFileList without error', () => {
    const wrapper = mount(<Upload {...defaultUploadProps}></Upload>)
    expect(wrapper).toMatchSnapshot()
  })

  // 7. test beforeUpload transformFile onChange can be called
  it('test callback', async () => {
    const fileList = [] as any
    const transformFileMock = jest.fn()
    const beforeUploadMock = jest.fn(() => {
      return true
    })
    const onChangeMock = jest.fn(() => {
      setTimeout(() => {
        expect(transformFileMock).toHaveBeenCalled()
      }, 100)
    })
    const wrapper = mount(
      <Upload
        action={action}
        transformFile={transformFileMock}
        fileList={fileList}
        beforeUpload={beforeUploadMock}
        onChange={onChangeMock}
      >
        {uploadButton}
      </Upload>,
    )
    const file = new File(['test'], 'test.png', { type: 'image/png' })
    wrapper.find('input').simulate('change', {
      target: {
        files: [file],
      },
    })
    expect(beforeUploadMock).toHaveBeenCalledWith(file, [file])
    await sleep(2000)
    expect(onChangeMock).toHaveBeenCalled()
  })

  // api
  it('api test', () => {
    const wrapper = mount(<Upload {...defaultUploadProps}>{uploadButton}</Upload>)

    wrapper.setProps({ showUploadList: false })
    expect(wrapper.find('.kd-upload-text-list').exists()).toBe(false)
    wrapper.setProps({ showUploadList: true })
    expect(wrapper.find('.kd-upload-text-list').exists()).toBe(true)
    expect(wrapper.find('.kd-upload-text-list-item-delete').exists()).toBe(true)
    expect(wrapper.find('.kd-upload-text-list-item-reupload').exists()).toBe(true)

    // accept
    wrapper.setProps({ accept: 'image/png, image/jpeg' })
    expect(wrapper.find('input').prop('accept')).toBe('image/png, image/jpeg')

    // disabled
    expect(wrapper.find('input').prop('disabled')).toBe(false)
    wrapper.setProps({ disabled: true })
    expect(wrapper.find('.kd-upload').hasClass('disabled')).toBe(true)
    expect(wrapper.find('input').prop('disabled')).toBe(true)

    // directory
    expect(wrapper.find('input').prop('directory')).toBe(undefined)
    wrapper.setProps({ directory: true })
    expect(wrapper.find('input').prop('directory')).toBe('directory')

    // multiple
    expect(wrapper.find('input').prop('multiple')).toBe(false)
    wrapper.setProps({ multiple: true })
    expect(wrapper.find('input').prop('multiple')).toBe(true)
  })

  it('onPreview', () => {
    const onPreview = jest.fn()
    const wrapper = mount(
      <Upload {...defaultUploadProps} onPreview={onPreview}>
        {uploadButton}
      </Upload>,
    )
    wrapper.find('.kd-upload-text-list-item-name').at(0).simulate('click')
    expect(onPreview).toHaveBeenCalled()
    wrapper.find('.kd-upload-text-list-item-icon').at(0).simulate('click')
    expect(onPreview).toHaveBeenCalledTimes(2)
  })

  it('onRemove', () => {
    let value = { ...defaultFileList[0] }
    const onRemove = jest.fn((v) => {
      value = v
    })
    const wrapper = mount(
      <Upload {...defaultUploadProps} onRemove={onRemove}>
        {uploadButton}
      </Upload>,
    )
    wrapper.find('.kd-upload-text-list-item-delete').at(0).simulate('click')
    expect(onRemove).toHaveBeenCalled()
    expect(value.status).toEqual('removed')
  })
})
