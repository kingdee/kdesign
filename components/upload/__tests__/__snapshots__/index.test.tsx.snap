// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload render defaultFileList without error 1`] = `
<Upload
  action="https://www.mocky.io/v2/5cc8019d300000980a055e76"
  defaultFileList={
    Array [
      Object {
        "name": "xxx.png",
        "response": "Server Error 500",
        "size": "100",
        "status": "done",
        "uid": "1",
        "url": "http://www.baidu.com/xxx.png",
      },
      Object {
        "name": "yyy.png",
        "size": "100",
        "status": "done",
        "uid": "2",
        "url": "http://www.baidu.com/yyy.png",
      },
      Object {
        "name": "zzz.png",
        "response": "Server Error 500",
        "size": "100",
        "status": "error",
        "uid": "3",
        "url": "http://www.baidu.com/zzz.png",
      },
    ]
  }
>
  <div
    className="kd-upload"
  >
    <ul
      className="kd-upload-text-list"
    >
      <Item
        disabled={false}
        file={
          Object {
            "name": "xxx.png",
            "response": "Server Error 500",
            "size": "100",
            "status": "done",
            "uid": "1",
            "url": "http://www.baidu.com/xxx.png",
          }
        }
        handleReUpload={[Function]}
        handleRemove={[Function]}
        key="1"
        listType="text"
        prefixCls="kd-upload"
      >
        <li
          className="kd-upload-text-list-item"
        >
          <span
            className="kd-upload-text-list-item-icon"
          >
            <Icon
              spin={false}
              type="attachment"
            >
              <i
                className="kdicon kdicon-attachment"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={Object {}}
              />
            </Icon>
          </span>
          <span
            className="kd-upload-text-list-item-name"
            title="xxx.png"
          >
            xxx.png
          </span>
          <span
            className="kd-upload-text-list-item-size"
          >
            (
            100bytes
            )
          </span>
          <div
            className="kd-upload-text-list-item-action"
          >
            <a
              className="kd-upload-text-list-item-delete"
              href="true"
              onClick={[Function]}
            >
              删除
            </a>
          </div>
        </li>
      </Item>
      <Item
        disabled={false}
        file={
          Object {
            "name": "yyy.png",
            "size": "100",
            "status": "done",
            "uid": "2",
            "url": "http://www.baidu.com/yyy.png",
          }
        }
        handleReUpload={[Function]}
        handleRemove={[Function]}
        key="2"
        listType="text"
        prefixCls="kd-upload"
      >
        <li
          className="kd-upload-text-list-item"
        >
          <span
            className="kd-upload-text-list-item-icon"
          >
            <Icon
              spin={false}
              type="attachment"
            >
              <i
                className="kdicon kdicon-attachment"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={Object {}}
              />
            </Icon>
          </span>
          <span
            className="kd-upload-text-list-item-name"
            title="yyy.png"
          >
            yyy.png
          </span>
          <span
            className="kd-upload-text-list-item-size"
          >
            (
            100bytes
            )
          </span>
          <div
            className="kd-upload-text-list-item-action"
          >
            <a
              className="kd-upload-text-list-item-delete"
              href="true"
              onClick={[Function]}
            >
              删除
            </a>
          </div>
        </li>
      </Item>
      <Item
        disabled={false}
        file={
          Object {
            "name": "zzz.png",
            "response": "Server Error 500",
            "size": "100",
            "status": "error",
            "uid": "3",
            "url": "http://www.baidu.com/zzz.png",
          }
        }
        handleReUpload={[Function]}
        handleRemove={[Function]}
        key="3"
        listType="text"
        prefixCls="kd-upload"
      >
        <li
          className="kd-upload-text-list-item error"
        >
          <span
            className="kd-upload-text-list-item-icon"
          >
            <Icon
              spin={false}
              type="warning-solid"
            >
              <i
                className="kdicon kdicon-warning-solid"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                style={Object {}}
              />
            </Icon>
          </span>
          <span
            className="kd-upload-text-list-item-name"
            title="zzz.png"
          >
            zzz.png
          </span>
          <span
            className="kd-upload-text-list-item-size"
          >
            (
            100bytes
            )
          </span>
          <div
            className="kd-upload-text-list-item-action"
          >
            <a
              className="kd-upload-text-list-item-reupload"
              href="true"
              onClick={[Function]}
            >
              重新上传
            </a>
            <a
              className="kd-upload-text-list-item-delete"
              href="true"
              onClick={[Function]}
            >
              删除
            </a>
          </div>
        </li>
      </Item>
    </ul>
  </div>
</Upload>
`;

exports[`Upload render no child without errors 1`] = `
<Upload>
  <div
    className="kd-upload"
  />
</Upload>
`;

exports[`Upload render null or undefined without errors 1`] = `
<Upload>
  <div
    className="kd-upload"
  />
</Upload>
`;

exports[`Upload renders correctly 1`] = `
<div
  class="kd-upload"
>
  <label
    class="kd-upload-drag"
  >
    <span
      class="kd-upload-handle"
    >
      <div>
        <i
          class="kdicon kdicon-add"
          style="font-size:16px;color:#666;font-weight:bolder"
        />
        <div>
          点击或拖拽上传
        </div>
      </div>
    </span>
    <span
      class="kd-upload-input"
    >
      <input
        title=""
        type="file"
      />
    </span>
  </label>
</div>
`;

exports[`Upload renders correctly 2`] = `
<div
  class="kd-upload"
/>
`;

exports[`Upload renders correctly 3`] = `
<div
  class="kd-upload"
>
  <ul
    class="kd-upload-picture-list"
  />
</div>
`;

